federation/pkg/federatedtypes/scheduling.go

```diff
@@ -87,6 +93,57 @@ func (a *replicaSchedulingAdapter) IsSchedulingAdapter() bool {
 	return true
 }
 
+func isSelected(names []string, name string) bool {
+	for _, val := range names {
+		if val == name {
+			return true
+		}
+	}
+	return false
+}
+
+func isObjHpaControlled(fedObj pkgruntime.Object) (bool, error) {
+	hpaSelectedClusters, error := hpautil.GetHpaTargetClusterList(fedObj)
+	if error != nil {
+		return false, error
+	}
+
+	if hpaSelectedClusters == nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+// initializeScheduleState initializes the schedule state for consumption by schedule
+// functions (schedule or simple schedule). After this initialization the state would
+// already have information, if only a subset of clusters targetted by hpa, or all clusters
+// need to be considered by the actual scheduling functions.
+func initializeScheduleState(fedObj pkgruntime.Object, clusterNames []string) (map[string]*ReplicaScheduleState, bool, error) {
+	initialState := make(map[string]*ReplicaScheduleState)
+	hpaControlled := false
+	hpaSelectedClusters, error := hpautil.GetHpaTargetClusterList(fedObj)
+	if error != nil {
+		return nil, hpaControlled, error
+	}
+
+	if hpaSelectedClusters != nil {
+		hpaControlled = true
+	}
+	for _, clusterName := range clusterNames {
+		replicaState := ReplicaScheduleState{
+			isSelected: false,
+			replicas:   0,
+		}
+		if hpaControlled {
+			if isSelected(hpaSelectedClusters.Names, clusterName) {
+				replicaState.isSelected = true
```

Comment:
ReplicaScheduleState stores the result of schedule() function specific to adapter (which cluster gets how many replicas). 0 means the particular cluster does not get any replicas, and is the initialised state too. Thought better to use 0  as that marker to avoid null checks. I have used an additional bool field to indicate if this cluster is selected in this reconcile pass or not, because we use the same field(s) to map the selection percolated from hpa or actual schedule function of this adapter.